File: static_files/assignments/HW2_files/SRC/MATLAB/gen_fx_MC_SVM.m
function [class, Mdl] = gen_fx_MC_SVM(sample, train, grp)
% gen_fx_MC_SVM - Multi-class SVM classification using one-vs-one strategy with voting.
%
% Syntax:
%   [class, Mdl] = gen_fx_MC_SVM(sample, train, grp)
%
% Inputs:
%   sample - Feature matrix of samples to classify (MxD)
%   train  - Feature matrix of training samples (NxD)
%   grp    - Class labels for the training samples (Nx1)
%
% Outputs:
%   class - Predicted class labels for the input samples (Mx1)
%   Mdl   - Trained ECOC (Error-Correcting Output Codes) model using SVM learners
%
% Description:
%   This function performs multi-class classification by constructing binary
%   SVM classifiers for each pair of classes (one-vs-one scheme). For each
%   sample, it uses a majority voting scheme to assign the final class.
%   Additionally, it trains an ECOC model as a compact representation for the final output.

classNo = double(unique(grp));    % Unique class labels
cls_nu = length(classNo);         % Number of classes

% Generate all possible binary class pairs (one-vs-one strategy)
r = nchoosek(1:cls_nu, 2);

vote = [];  % Will hold votes from all pairwise classifiers

% Loop through each class pair and train a binary SVM
for i = 1:length(r(:,1))
    % Select samples from only the two classes
    ix = (grp == classNo(r(i,1))) | (grp == classNo(r(i,2)));
    
    % Train binary SVM on selected samples
    svmStruct = fitcsvm(train(ix,:), grp(ix));
    
    % Predict class for all test samples and collect the votes
    vote = [vote predict(svmStruct, sample)];
end

% Determine final predicted class by majority vote
class = mode(vote, 2);

% Train a multi-class SVM using ECOC as a model output (not used in prediction here)
Mdl = fitcecoc(train, grp);
% Optionally, one could use:
% class = predict(Mdl, sample);
% to classify using the ECOC model directly instead of voting
end

File: static_files/assignments/HW2_files/SRC/Python/gen_fx_MC_SVM.py

import numpy as np
from sklearn.svm import SVC
from sklearn.multiclass import OneVsOneClassifier
from scipy.stats import mode

def gen_fx_MC_SVM(sample, train, grp):
    """
    Multi-class SVM classification using one-vs-one strategy with voting.

    Parameters:
        sample (np.ndarray): Feature matrix of samples to classify (MxD)
        train (np.ndarray): Feature matrix of training samples (NxD)
        grp (np.ndarray): Class labels for the training samples (Nx1 or 1D)

    Returns:
        class_pred (np.ndarray): Predicted class labels for the input samples (Mx1)
        model (OneVsOneClassifier): Trained One-vs-One multi-class SVM model
    """
    grp = np.array(grp).flatten()
    classNo = np.unique(grp)
    cls_nu = len(classNo)

    # Store predictions from all pairwise classifiers
    vote_matrix = np.zeros((sample.shape[0], cls_nu * (cls_nu - 1) // 2))
    
    idx = 0
    for i in range(cls_nu):
        for j in range(i + 1, cls_nu):
            class_i = classNo[i]
            class_j = classNo[j]

            # Select samples from the two classes
            ix = np.where((grp == class_i) | (grp == class_j))[0]
            X_pair = train[ix]
            y_pair = grp[ix]

            # Train binary SVM
            clf = SVC(kernel='linear')
            clf.fit(X_pair, y_pair)

            # Predict test samples
            vote_matrix[:, idx] = clf.predict(sample)
            idx += 1

    # Majority voting
    class_pred, _ = mode(vote_matrix, axis=1)
    class_pred = class_pred.flatten()

    # Train ECOC-style model (sklearn uses OvO under the hood)
    model = OneVsOneClassifier(SVC(kernel='linear'))
    model.fit(train, grp)

    return class_pred, model
